File: AspNetMvcUnleashed/0672329980_source/Chapter11/VB/MvcApplication1/MvcApplication1/Controllers/SlidingController.vb
Public Class SlidingController
    Inherits Controller

    Private _entities As New MoviesDBEntities()

    Public Function Details(ByVal id As Integer) As ActionResult
        Dim cache = Me.HttpContext.Cache
        Dim key = GetMovieCacheKey(id)
        Dim movie = CType(cache(key), Movie)


        If movie IsNot Nothing Then
            Debug.WriteLine("Got movie from cache")
        Else
            Debug.WriteLine("Getting movie from database")
            movie = (From m In _entities.MovieSet _
                     Where m.Id = id _
                     Select m).FirstOrDefault()
            cache.Insert(key, movie, Nothing, cache.NoAbsoluteExpiration, TimeSpan.FromMinutes(10))
        End If

        Return View(movie)
    End Function


    Private Function GetMovieCacheKey(ByVal movieId As Integer) As String
        Return "movie" & movieId.ToString()
    End Function

End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter11/CS/MvcApplication1/MvcApplication1/Controllers/SlidingController.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Web.Caching;
using System.Web.Mvc;
using MvcApplication1.Models;

namespace MvcApplication1.Controllers
{
    public class SlidingController : Controller
    {
        private MoviesDBEntities _entities = new MoviesDBEntities(); 

        public ActionResult Details(int id)
        {
            var cache = this.HttpContext.Cache;
            var key = GetMovieCacheKey(id);
            var movie = (Movie)cache[key];


            if (movie != null)
            {
                Debug.WriteLine("Got movie from cache");
            }
            else
            {
                Debug.WriteLine("Getting movie from database");
                movie = (from m in _entities.MovieSet
                         where m.Id == id
                         select m).FirstOrDefault();
                cache.Insert(key, movie, null, Cache.NoAbsoluteExpiration, TimeSpan.FromMinutes(10));
            }

            return View(movie);
        }


        private string GetMovieCacheKey(int movieId)
        {
            return "movie" + movieId.ToString();
        }
    }
}
